casper/src/main/scala/coop/rchain/casper/Casper.scala

```diff
@@ -177,8 +183,54 @@ sealed abstract class MultiParentCasperInstances {
           tip       = estimates.head
           _ <- Log[F].info(
                 s"CASPER: New fork-choice tip is block ${PrettyPrinter.buildString(tip.blockHash)}.")
+          lastFinalizedBlock        <- lastFinalizedBlockContainer.get
+          updatedLastFinalizedBlock <- updateLastFinalizedBlock(dag, lastFinalizedBlock)
+          _                         <- lastFinalizedBlockContainer.set(updatedLastFinalizedBlock)
         } yield attempt
 
+      def updateLastFinalizedBlock(dag: BlockDag,
+                                   lastFinalizedBlock: BlockMessage): F[BlockMessage] = {
+        val maybeFinalizedBlockChildren = dag.childMap.get(lastFinalizedBlock.blockHash)
+        maybeFinalizedBlockChildren match {
+          case Some(finalizedBlockChildren) =>
+            updateLastFinalizedBlockAux(dag, lastFinalizedBlock, finalizedBlockChildren.toSeq)
+          case None => lastFinalizedBlock.pure[F]
+        }
+      }
+
+      def updateLastFinalizedBlockAux(
+          dag: BlockDag,
+          lastFinalizedBlock: BlockMessage,
+          finalizedBlockCandidatesHashes: Seq[BlockHash]): F[BlockMessage] =
+        finalizedBlockCandidatesHashes match {
+          case Nil => lastFinalizedBlock.pure[F]
+          case blockHash +: rem =>
+            for {
+              maybeBlock <- BlockStore[F].get(blockHash)
+              updatedLastFinalizedBlock <- maybeBlock match {
+                                            case Some(block) =>
+                                              for {
+                                                normalizedFaultTolerance <- SafetyOracle[F]
+                                                                             .normalizedFaultTolerance(
+                                                                               dag,
+                                                                               block)
+                                                updatedLastFinalizedBlock <- if (normalizedFaultTolerance > faultToleranceThreshold) {
+                                                                              updateLastFinalizedBlock(
```

Comment:
I don't think this algorithm is right. It is presently a depth first search because you proceed down the path of the first finalized child. But you want a breadth-first search to ensure you actually go as far as possible. This is why I was trying to update `bfTraverse` to take an `F`, but I have not been successful yet. This is why the PR is blocked.

Reply:
As per earlier discussion with Kent, my understanding was incorrect. The algorithm here is correct because at most one child per block can be above the fault tolerance threshold.